src/NLog/Targets/TargetWithContext.cs

```diff
@@ -286,33 +286,48 @@ protected bool ShouldIncludeProperties(LogEventInfo logEvent)
         {
             if (EstimateAllPropertiesCount(logEvent, out var logEventProperties, out var scopeContextProperties, out var gdcNames, out int propertiesCount) > 0)
             {
-                List<KeyValuePair<string, object>> combinedProperties = new List<KeyValuePair<string, object>>(propertiesCount);
                 bool checkExcludeProperties = ExcludeProperties.Count > 0;
-
-                Action<string, object> tryAddProperty = (string propertyName, object propertyValue) =>
+                Func<string, object, bool, KeyValuePair<string, object> ?> buildPair = (string propertyName, object propertyValue, bool serialize) =>
                 {
                     if (string.IsNullOrEmpty(propertyName))
-                        return;
+                        return null;
 
                     if (checkExcludeProperties && ExcludeProperties.Contains(propertyName))
-                        return;
+                        return null;
 
-                    if (SerializeScopeContextProperty(logEvent, propertyName, propertyValue, out var serializedValue))
+                    if(serialize)
                     {
-                        combinedProperties.Add(new KeyValuePair<string, object>(propertyName, propertyValue));
+                        if(SerializeScopeContextProperty(logEvent, propertyName, propertyValue, out var serializedValue))
+                        {
+                            return new KeyValuePair<string, object>(propertyName, serializedValue);
+                        }
+                        else
+                        {
+                            return null;
+                        }
                     }
+
+                    return new KeyValuePair<string, object>(propertyName, propertyValue);
                 };
 
-                if(logEventProperties?.Count > 0)
+                if (logEventProperties?.Count > 0)
                 {
                     foreach (var property in logEventProperties)
-                        tryAddProperty(property.Key.ToString(), property.Value);
+                    {
+                        var pair = buildPair(property.Key.ToString(), property.Value, false);
+                        if (pair != null)
+                            yield return pair.Value;
```

Comment:
Can the yield-engine be moved into its own method, so the yield-engine is only allocated when something to yield. Move the yield-engine into a private method, and ensure `GetAllPropertiesList` only calls the private method when `EstimateAllPropertiesCount` > 0.

Right now the yield-engine will be allocated when entering `GetAllPropertiesList`, because the method now contains the `yield`-keyword.

Reply:
I've removed `EstimateAllPropertiesCount()` method, and introduced `GetAllPropertiesCountContext` class.

Prev. `EstimateAllPropertiesCount` is now `ctx.PropertiesCount`

yield-engine moved into `GetAllPropertiesCountContext.BuildList()`

IMHO, now the code looks more elegant.